src/runtime/device_interface.cpp

```diff
@@ -387,4 +390,176 @@ WEAK void halide_device_and_host_free_as_destructor(void *user_context, void *ob
+        // OpenGL texture). If the dst device interface doesn't
+        // recognize the src device interface, it should ask the src
+        // device interface to copy to host memory in the *dst* buffer
+        // first (so that only the subset required is copied), and
+        // then do a copy_to_device from there.
+        err = dst_device_interface->impl->buffer_copy(user_context, src, dst_device_interface, dst);
+    } else if (src_device_interface) {
+        // The src device interface can handle copies to host
+        err = src_device_interface->impl->buffer_copy(user_context, src, dst_device_interface, dst);
+    } else {
+        // This is a host->host copy. The default implementation can
+        // handle this.
+        err = halide_default_buffer_copy(user_context, src, dst_device_interface, dst);
+    }
+
+    if (dst != src) {
+        if (dst_device_interface) {
+            dst->set_device_dirty(true);
+        } else {
+            dst->set_host_dirty(true);
+        }
+    }
+
+    if (err) {
+        err = halide_error_code_device_buffer_copy_failed;
+    }
+
+    if (dst_device_interface) {
+        dst_device_interface->impl->release_module();
+    }
+    if (src_device_interface) {
+        src_device_interface->impl->release_module();
+    }
+
+    return err;
+}
+
+
+WEAK int halide_default_device_crop(void *user_context,
+                                    const struct halide_buffer_t *src,
+                                    struct halide_buffer_t *dst) {
+    error(user_context) << "device_interface does not support cropping\n";
+    return halide_error_code_device_crop_unsupported;
+}
+
+WEAK int halide_device_crop(void *user_context,
+                            const struct halide_buffer_t *src,
+                            struct halide_buffer_t *dst) {
+    ScopedMutexLock lock(&device_copy_mutex);
+
+    if (!src->device) {
+        return 0;
+    }
+
+    if (dst->device) {
+        error(user_context) << "destination buffer already has a device allocation\n";
+        return halide_error_code_device_crop_failed;
+    }
+
+    src->device_interface->impl->use_module();
+    int err = src->device_interface->impl->device_crop(user_context, src, dst);
+
+    debug(user_context) << "halide_device_crop " << "\n"
+                        << " src: " << *src << "\n"
+                        << " dst: " << *dst << "\n";
+
+    return err;
+}
+
+WEAK int halide_default_device_release_crop(void *user_context,
+                                            struct halide_buffer_t *buf) {
+    if (!buf->device) {
+        return 0;
+    }
+    error(user_context) << "device_interface does not support cropping\n";
+    return halide_error_code_device_crop_unsupported;
+}
+
+
+WEAK int halide_device_release_crop(void *user_context,
+                                    struct halide_buffer_t *buf) {
+    if (!buf->device) {
+        return 0;
+    }
+
+    {
```

Comment:
Stylistically, I like putting scoped mutex locks at the top of a scope, so that it's clear to the reader what the mutex encompasses.